### PR TITLE
Add flatScheduledTask() methods to EventLoop

### PR DESCRIPTION
**Pull request by gwynne**
Add `flatScheduleTask(deadline:)` and `flatScheduleTask(in:)` methods to `EventLoop`.

### Motivation:

Currently the best, or at least simplest, way to schedule a delayed asynchronous task and wait on its complete result is something like `eventLoop.scheduleTask(in: ...) { return somethingAsync(...) }.futureResult.flatMap { $0 }`. This is not only unobvious but also a bit ugly.

### Modifications:

I have added methods which correspond to scheduleTask() in the same way flatSubmit() does to submit(). The returned Scheduled<T> object allows cancellation of the asynchronous task's execution in the same way that is possible for synchronous scheduled tasks, and the futureResult does not complete until the entire asynchronous task does. Callers wishing to wait only for submission of the delayed task, rather than its overall completion, may continue to use the existing scheduleTask() APIs to do so. Tests have been added to verify that the behavior of the new methods is as expected.

I also made a few very minor corrections to a couple of documentation comments which suffered from copy-pasta.

### Result:

These changes are purely additive.

### CODE REVIEW

File: Sources/NIO/EventLoop.swift [hunk 1]
```diff
@@ -482,13 +482,43 @@ extension EventLoop {
...
+    ///     - task: The asynchronous task to run. As everything that runs on the `EventLoop`, it must not block.
     /// - returns: An `EventLoopFuture` identical to the `EventLooopFuture` returned from `task`.
     @inlinable
     public func flatSubmit<T>(_ task: @escaping () -> EventLoopFuture<T>) -> EventLoopFuture<T> {
         return self.submit(task).flatMap { $0 }
     }
 
+    /// Schedule a `task` that is executed by this `EventLoop` at the given time.
+    ///
+    /// - parameters:
```
Comment (weissi) [inline Sources/NIO/EventLoop.swift:494]: may be worth adding an

```
/// - info: You can only cancel a task before it has started executing.
```

File: Tests/NIOTests/EventLoopTest.swift [hunk 2]
```diff
@@ -139,6 +179,22 @@ public final class EventLoopTest : XCTestCase {
...
+        }
+
+        let loop = eventLoopGroup.next()
+        loop.execute {
+            let task = loop.flatScheduleTask(in: .milliseconds(0)) {
+                loop.makeSucceededFuture(()).map { XCTFail() }
+            }
+            task.cancel()
+        }
+        Thread.sleep(until: .init(timeIntervalSinceNow: 0.1))
```
[the remark this answers is not in the document]
Comment (Lukasa) [reply Tests/NIOTests/EventLoopTest.swift:195]: I assume this is to wait for confirmation that the scheduled task could have run. The better way to test this, I think, would be to schedule a task that would be _required_ to run after the cancelled task, and wait for that to complete. Then we know it must have either completed or never run.

File: Sources/NIO/EventLoop.swift [hunk 3]
```diff
@@ -482,13 +482,43 @@ extension EventLoop {
...
+    ///     - task: The asynchronous task to run. As everything that runs on the `EventLoop`, it must not block.
+    /// - returns: A `Scheduled` object which may be used to cancel the task if it has not yet run, or to wait
+    ///            on the full execution of the task, including its returned `EventLoopFuture`.
+    @discardableResult
+    func flatScheduleTask<T>(deadline: NIODeadline, _ task: @escaping () throws -> EventLoopFuture<T>) -> Scheduled<T> {
+        let promise: EventLoopPromise<T> = makePromise(file: #file, line: #line)
```
Comment (glbrntt) [inline Sources/NIO/EventLoop.swift:500]: nit: explicit self

File: Sources/NIO/EventLoop.swift [hunk 4]
```diff
@@ -482,13 +482,43 @@ extension EventLoop {
...
     }
 
+    /// Schedule a `task` that is executed by this `EventLoop` at the given time.
+    ///
+    /// - parameters:
+    ///     - task: The asynchronous task to run. As everything that runs on the `EventLoop`, it must not block.
+    /// - returns: A `Scheduled` object which may be used to cancel the task if it has not yet run, or to wait
+    ///            on the full execution of the task, including its returned `EventLoopFuture`.
+    @discardableResult
+    func flatScheduleTask<T>(deadline: NIODeadline, _ task: @escaping () throws -> EventLoopFuture<T>) -> Scheduled<T> {
```
Comment (glbrntt) [inline Sources/NIO/EventLoop.swift:499]: I assume this is meant to be `public`!

File: Sources/NIO/EventLoop.swift [hunk 5]
```diff
@@ -482,13 +482,43 @@ extension EventLoop {
...
+    }
+
+    /// Schedule a `task` that is executed by this `EventLoop` after the given amount of time.
+    ///
+    /// - parameters:
+    ///     - task: The asynchronous task to run. As everything that runs on the `EventLoop`, it must not block.
+    /// - returns: A `Scheduled` object which may be used to cancel the task if it has not yet run, or to wait
+    ///            on the full execution of the task, including its returned `EventLoopFuture`.
+    @discardableResult
+    func flatScheduleTask<T>(in delay: TimeAmount, _ task: @escaping () throws -> EventLoopFuture<T>) -> Scheduled<T> {
```
Comment (glbrntt) [inline Sources/NIO/EventLoop.swift:514]: Same as above re: public and explicit self

File: Tests/NIOTests/EventLoopTest.swift [hunk 6]
```diff
@@ -91,6 +106,31 @@ public final class EventLoopTest : XCTestCase {
         XCTAssertFalse(ran.load())
     }
 
+    public func testFlatScheduleCancelled() throws {
+        let eventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
```
[the remark this answers is not in the document]
Comment (gwynne) [reply Tests/NIOTests/EventLoopTest.swift]: Done!

File: Tests/NIOTests/EventLoopTest.swift [hunk 7]
```diff
@@ -33,6 +33,21 @@ public final class EventLoopTest : XCTestCase {
         XCTAssertTrue(value)
     }
 
+    public func testFlatSchedule() throws {
+        let nanos: NIODeadline = .now()
+        let amount: TimeAmount = .seconds(1)
+        let eventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
```
[the remark this answers is not in the document]
Comment (gwynne) [reply Tests/NIOTests/EventLoopTest.swift]: Done!

File: Tests/NIOTests/EventLoopTest.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (gwynne) [reply Tests/NIOTests/EventLoopTest.swift]: Rewritten using EmbeddedEventLoop per @weissi 's suggestions

File: Sources/NIO/EventLoop.swift (hunk 1 above)
Comment (gwynne) [reply Sources/NIO/EventLoop.swift:508]: Done!

File: Sources/NIO/EventLoop.swift (hunk 3 above)
Comment (gwynne) [reply Sources/NIO/EventLoop.swift]: Fixed!

File: Sources/NIO/EventLoop.swift (hunk 4 above)
Comment (gwynne) [reply Sources/NIO/EventLoop.swift]: Yep, whoopsy! That's a fun one, make some new APIs that no one can actually use! 😆 Fixed!

File: Sources/NIO/EventLoop.swift [hunk 8]
```diff
@@ -236,11 +236,25 @@ public protocol EventLoop: EventLoopGroup {
     /// - returns: `EventLoopFuture` that is notified once the task was executed.
     func submit<T>(_ task: @escaping () throws -> T) -> EventLoopFuture<T>
 
-    /// Schedule a `task` that is executed by this `SelectableEventLoop` at the given time.
+    /// Schedule a `task` that is executed by this `EventLoop` at the given time.
+    ///
+    /// - parameters:
+    ///     - task: The synchronous task to run. As everything that runs on the `EventLoop`, it must not block.
```
Comment (Lukasa) [inline Sources/NIO/EventLoop.swift:242]:
```suggestion
    ///     - task: The synchronous task to run. As with everything that runs on the `EventLoop`, it must not block.
```

File: Sources/NIO/EventLoop.swift [hunk 9]
```diff
@@ -236,11 +236,25 @@ public protocol EventLoop: EventLoopGroup {
...
+    ///
+    /// - note: You can only cancel a task before it has started executing.
     @discardableResult
     func scheduleTask<T>(deadline: NIODeadline, _ task: @escaping () throws -> T) -> Scheduled<T>
 
-    /// Schedule a `task` that is executed by this `SelectableEventLoop` after the given amount of time.
+    /// Schedule a `task` that is executed by this `EventLoop` after the given amount of time.
+    ///
+    /// - parameters:
+    ///     - task: The synchronous task to run. As everything that runs on the `EventLoop`, it must not block.
```
Comment (Lukasa) [inline Sources/NIO/EventLoop.swift:253]:
```suggestion
    ///     - task: The synchronous task to run. As with everything that runs on the `EventLoop`, it must not block.
```

File: Sources/NIO/EventLoop.swift [hunk 10]
```diff
@@ -482,13 +496,47 @@ extension EventLoop {
     /// the `EventLoopFuture` returned by `task`.
     ///
     /// - parameters:
-    ///     - task: The synchronous task to run. As everything that runs on the `EventLoop`, it must not block.
+    ///     - task: The asynchronous task to run. As everything that runs on the `EventLoop`, it must not block.
```
Comment (Lukasa) [inline Sources/NIO/EventLoop.swift:499]:
```suggestion
    ///     - task: The asynchronous task to run. As with everything that runs on the `EventLoop`, it must not block.
```

File: Sources/NIO/EventLoop.swift [hunk 11]
```diff
@@ -482,13 +496,47 @@ extension EventLoop {
...
     /// - returns: An `EventLoopFuture` identical to the `EventLooopFuture` returned from `task`.
     @inlinable
     public func flatSubmit<T>(_ task: @escaping () -> EventLoopFuture<T>) -> EventLoopFuture<T> {
         return self.submit(task).flatMap { $0 }
     }
 
+    /// Schedule a `task` that is executed by this `EventLoop` at the given time.
+    ///
+    /// - parameters:
+    ///     - task: The asynchronous task to run. As everything that runs on the `EventLoop`, it must not block.
```
Comment (Lukasa) [inline Sources/NIO/EventLoop.swift:509]:
```suggestion
    ///     - task: The asynchronous task to run. As with everything that runs on the `EventLoop`, it must not block.
```

File: Tests/NIOTests/EventLoopTest.swift [hunk 12]
```diff
@@ -19,18 +19,45 @@ import NIOConcurrencyHelpers
...
+
+        var result: Bool?
+        scheduled.futureResult.whenSuccess { result = $0 }
+        
+        eventLoop.run() // run without time advancing should do nothing
+        XCTAssertFalse(scheduled.futureResult.isFulfilled)
+        XCTAssertNil(result)
+        
+        eventLoop.advanceTime(by: .seconds(1))
+        eventLoop.run() // should fire now
```
Comment (Lukasa) [inline Tests/NIOTests/EventLoopTest.swift:34]: `advanceTime` implies `run`, so we shouldn't need both.

File: Tests/NIOTests/EventLoopTest.swift [hunk 13]
```diff
@@ -19,18 +19,45 @@ import NIOConcurrencyHelpers
...
```
Comment (Lukasa) [inline Tests/NIOTests/EventLoopTest.swift:56]: Same note, `advanceTime` implies `run`.

File: Tests/NIOTests/EventLoopTest.swift [hunk 14]
```diff
@@ -69,26 +96,51 @@ public final class EventLoopTest : XCTestCase {
...
+
+        let scheduled = eventLoop.scheduleTask(in: .seconds(1)) { true }
+
+        var result: Bool?
+        var error: Error?
+        scheduled.futureResult.whenSuccess { result = $0 }
+        scheduled.futureResult.whenFailure { error = $0 }
+        
+        eventLoop.advanceTime(by: .milliseconds(500))
+        eventLoop.run() // advance halfway to firing time
```
Comment (Lukasa) [inline Tests/NIOTests/EventLoopTest.swift:109]: Same note, `advanceTime` implies `run`.

File: Tests/NIOTests/EventLoopTest.swift [hunk 15]
```diff
@@ -69,26 +96,51 @@ public final class EventLoopTest : XCTestCase {
...
 
         scheduled.cancel()
+        
+        eventLoop.advanceTime(by: .milliseconds(500))
+        eventLoop.run() // advance the rest of the way
```
Comment (Lukasa) [inline Tests/NIOTests/EventLoopTest.swift:114]: And here.

File: Tests/NIOTests/EventLoopTest.swift [hunk 16]
```diff
@@ -69,26 +96,51 @@ public final class EventLoopTest : XCTestCase {
...
+            eventLoop.makeSucceededFuture(true)
+        }
+
+        var result: Bool?
+        var error: Error?
+        scheduled.futureResult.whenSuccess { result = $0 }
+        scheduled.futureResult.whenFailure { error = $0 }
+        
+        eventLoop.advanceTime(by: .milliseconds(500))
+        eventLoop.run() // advance halfway to firing time
```
Comment (Lukasa) [inline Tests/NIOTests/EventLoopTest.swift:134]: And here.

File: Tests/NIOTests/EventLoopTest.swift [hunk 17]
```diff
@@ -69,26 +96,51 @@ public final class EventLoopTest : XCTestCase {
...
+
+        scheduled.cancel()
+        
+        eventLoop.advanceTime(by: .milliseconds(500))
+        eventLoop.run() // advance the rest of the way
```
Comment (Lukasa) [inline Tests/NIOTests/EventLoopTest.swift:139]: And here.

File: Tests/NIOTests/EventLoopTest.swift [hunk 18]
```diff
@@ -124,19 +176,41 @@ public final class EventLoopTest : XCTestCase {
...
-            }
-            task.cancel()
+        var result: Bool?
+        var error: Error?
+        scheduled.futureResult.whenSuccess { result = $0 }
+        scheduled.futureResult.whenFailure { error = $0 }
+        
+        scheduled.cancel()
+        eventLoop.advanceTime(by: .seconds(1))
+        eventLoop.run()
```
Comment (Lukasa) [inline Tests/NIOTests/EventLoopTest.swift:189]: and here.

File: Tests/NIOTests/EventLoopTest.swift [hunk 19]
```diff
@@ -124,19 +176,41 @@ public final class EventLoopTest : XCTestCase {
...
-        Thread.sleep(until: .init(timeIntervalSinceNow: 0.1))
+
+        var result: Bool?
+        var error: Error?
+        scheduled.futureResult.whenSuccess { result = $0 }
+        scheduled.futureResult.whenFailure { error = $0 }
+        
+        scheduled.cancel()
+        eventLoop.advanceTime(by: .seconds(1))
+        eventLoop.run()
```
Comment (Lukasa) [inline Tests/NIOTests/EventLoopTest.swift:209]: And here.

File: Sources/NIO/EventLoop.swift [hunk 20]
```diff
@@ -482,13 +496,47 @@ extension EventLoop {
...
+    /// Schedule a `task` that is executed by this `EventLoop` at the given time.
+    ///
+    /// - parameters:
+    ///     - task: The asynchronous task to run. As with everything that runs on the `EventLoop`, it must not block.
+    /// - returns: A `Scheduled` object which may be used to cancel the task if it has not yet run, or to wait
+    ///            on the full execution of the task, including its returned `EventLoopFuture`.
+    ///
+    /// - note: You can only cancel a task before it has started executing.
+    @discardableResult
+    public func flatScheduleTask<T>(deadline: NIODeadline, _ task: @escaping () throws -> EventLoopFuture<T>) -> Scheduled<T> {
```
Comment (Lukasa) [inline Sources/NIO/EventLoop.swift:515]: Can we make `file` and `line` parameters of this function and then pass them to the line below? That will be much more useful, debugging-wise.

File: Sources/NIO/EventLoop.swift [hunk 21]
```diff
@@ -482,13 +496,47 @@ extension EventLoop {
...
+    /// Schedule a `task` that is executed by this `EventLoop` after the given amount of time.
+    ///
+    /// - parameters:
+    ///     - task: The asynchronous task to run. As everything that runs on the `EventLoop`, it must not block.
+    /// - returns: A `Scheduled` object which may be used to cancel the task if it has not yet run, or to wait
+    ///            on the full execution of the task, including its returned `EventLoopFuture`.
+    ///
+    /// - note: You can only cancel a task before it has started executing.
+    @discardableResult
+    public func flatScheduleTask<T>(in delay: TimeAmount, _ task: @escaping () throws -> EventLoopFuture<T>) -> Scheduled<T> {
```
Comment (Lukasa) [inline Sources/NIO/EventLoop.swift:532]: Same file/line note here.

File: Sources/NIO/EventLoop.swift (hunk 20 above)
Comment (gwynne) [reply Sources/NIO/EventLoop.swift]: Done,

File: Sources/NIO/EventLoop.swift (hunk 21 above)
Comment (gwynne) [reply Sources/NIO/EventLoop.swift]: and done 🙂 I put the file and line parameters between the time parameter and the closure so trailing closure syntax would continue to be available.

**Pull request merged**